src/blr/cli/tagbam.py

```diff
@@ -1,58 +1,100 @@
-                name, raw_barcode_tag, corr_barcode_tag = read.query_name.split()[0], None, None
+            for tag in args.tags:
 
-            # Rename aligment to exclude barcode information.
-            read.query_name = name
+                # Make regex expression and search for tag
+                pattern = regex_function(bam_tag=tag)
+                match = re.search(pattern, read.query_name)
 
-            if corr_barcode_tag:
-                assert corr_barcode_tag.startswith(f"{args.barcode_tag}:Z:")
-                corr_barcode = corr_barcode_tag.split(":")[-1]
-                read.set_tag(args.barcode_tag, corr_barcode, value_type='Z')
-
-            if raw_barcode_tag:
-                assert raw_barcode_tag.startswith(f"{args.sequence_tag}:Z:")
-                raw_barcode = raw_barcode_tag.split(":")[-1]
-                read.set_tag(args.sequence_tag, raw_barcode, value_type='Z')
+                # If tag string is found, remove from header and set SAM tag value
+                if match:
+                    full_tag_string = match.group(0)
+                    divider = read.query_name[match.start() - 1]
+                    read.query_name = read.query_name.replace(divider + full_tag_string, "")
+                    if not args.only_remove:
+                        read.set_tag(tag, match.group("value"), value_type=match.group("type"))
+                    summary[f"reads with tag {tag}"] += 1
+                else:
+                    summary[f"reads without tag {tag}"] += 1
 
             out.write(read)
 
-    logger.info(f"Alignments missing barcodes: {alignments_missing_bc}")
+    print_stats(summary, name="stats")
     logger.info("Finished")
 
 
+def build_regex_sam_tag(bam_tag, allowed_value_chars="ATGCN"):
+    """
+    Buidls regex string for SAM tags.
+    :param bam_tag: str, SAM tag to search for, e.g. BX
+    :param allowed_value_chars: str, characters allowed in SAM value
+    :return: str, regex expression of a SAM tag
+    """
+
+    # Build regex pattern strings
+    pattern_tag_value = f"[{allowed_value_chars}]+"
+    pattern_tag_types = f"[{ALLOWED_SAM_TAG_TYPES}]"
+
+    # Add strings to name match object variables to match.group(<name>)
+    regex_bam_tag = add_regex_name(pattern=bam_tag, name="tag")
+    regex_allowed_types = add_regex_name(pattern=pattern_tag_types, name="type")
+    regex_tag_value = add_regex_name(pattern=pattern_tag_value, name="value")
+
+    # regex pattern search for tag:type:value
+    regex_string = r":".join([regex_bam_tag, regex_allowed_types, regex_tag_value])
+    return regex_string
+
+
+def add_regex_name(pattern, name):
+    """
+    Formats a string to fit the regex pattern for getting named match objects groups
+    :param pattern: str, regex pattern
+    :param name: str, name of the match group
+    :return: str, named regex string
+    """
+    prefix = "(?P<"
+    infix = ">"
+    suffix = ")"
+    return prefix + name + infix + pattern + suffix
```

Comment:
```suggestion
    return "".join([prefix, name, infix, pattern, suffix])
```
`join` is faster then adding strings together. When you add strings it sets the new string for each addition i.e 
first `prefix + name`, 
then `prefix + name + infix`, 
then `prefix + name + infix + pattern`, 
then `prefix + name + infix + pattern + suffix`

Reply:
Cool, had no idea!

I'll also include a reference look for method 6 in this [investigation](https://waymoot.org/home/python_string/).

Reply:
The regular expression is created only once, and the runtime is therefore negligible compared to the other work this script does. Also, `''.join()` may not actually be faster because to use it you need to build a list first.

Base your decision on what to use here only on what you find more readable.